File: hestiaRUST/hestia_testing/log_test.rs
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.
use crate::hestia_testing;
use crate::hestia_testing::testlibs_test;

// test suites
const SUITE_NAME: &str = "hestia_testing::log API";

// test libs
fn assert_panic(s: &hestia_testing::Scenario) -> bool {
	if hestia_testing::has_condition(s, testlibs_test::COND_PROPER_LOG_DATA) {
		return true;
	}

	return false;
}

fn assert_log(s: &hestia_testing::Scenario, ts: &hestia_testing::Scenario) -> bool {
	if hestia_testing::has_condition(s, testlibs_test::COND_PROPER_LOG_DATA) {
		if ts.logs.iter().any(|i| i == testlibs_test::VALUE_LOG_DATA) {
			return true;
		}
	}

	return false;
}

fn test_log_algorithm(id: u64, desc: String, switches: Vec<String>) {
	// prepare
	let mut s: &mut hestia_testing::Scenario = &mut hestia_testing::new_scenario();
	s.id = id;
	s.name = String::from(SUITE_NAME);
	s.description = desc;
	s.switches = switches;

	// test
	let ts: &mut hestia_testing::Scenario = &mut hestia_testing::new_scenario();

	let statement = testlibs_test::create_statement(s);
	hestia_testing::log(s, format!("Given string	: {}\n", statement));

	hestia_testing::conclude(s, hestia_testing::VERDICT_PASS);
	if hestia_testing::has_condition(s, testlibs_test::COND_EMPTY_LOG_DATA) {
		println!("{}", hestia_testing::to_string(s));
	}

	hestia_testing::log(ts, statement);

	// assert
	if !assert_panic(&s) {
		hestia_testing::conclude(s, hestia_testing::VERDICT_FAIL);
	}

	if !assert_log(&s, &ts) {
		hestia_testing::conclude(s, hestia_testing::VERDICT_FAIL);
	}

	// report
	if !hestia_testing::has_condition(s, testlibs_test::COND_EMPTY_LOG_DATA) {
		println!("{}", hestia_testing::to_string(s));
	}
	assert!(hestia_testing::conclusion(s) == hestia_testing::VERDICT_PASS);
}

// test suites
hestia_testing_exec!(log_empty_statement, true, {
	test_log_algorithm(
		1,
		"\
Test Log() is able to panic when an empty string statement is provided.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_EMPTY_LOG_DATA)],
	)
});

hestia_testing_exec!(log_proper_statement, {
	test_log_algorithm(
		0,
		"\
Test Log() is able to process properly when a proper string statement is provided.
"
		.to_string(),
		vec![String::from(testlibs_test::COND_PROPER_LOG_DATA)],
	)
});

File: hestiaGO/hestiaTESTING/Log_test.go
// Copyright 2022 "Holloway" Chew, Kean Ho <kean.ho.chew@zoralab.com>
// Copyright 2022 ZORALab Enterprise <tech@zoralab.com>
//
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not
// use this file except in compliance with the License. You may obtain a copy of
// the License at
//
//                  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations under
// the License.

package hestiaTESTING

import (
	"testing"
)

func _testLogScenarios() []*Scenario {
	return []*Scenario{
		{
			Description: `
Test Log() is able to work properly with proper Scenario settings.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to work properly with empty Name setting.
`,
			Switches: []string{
				cond_EMPTY_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to work properly with empty Switches setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_EMPTY_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to work properly with nil Switches setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_NIL_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to work properly with empty log setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_EMPTY_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to work properly with nil log setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_NIL_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to work properly with empty description setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_EMPTY_DESCRIPTION,
				cond_NIL_LOG,
				cond_PROPER_SWITCHES,
				cond_PROPER_STRING_STATEMENT,
			},
		}, {
			Description: `
Test Log() is able to panic when nil Scenario is supplied.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_SUPPLY_NIL_SCENARIO,
				cond_PROPER_STRING_STATEMENT,
				expect_PANIC,
			},
		}, {
			Description: `
Test Log() is able to panic with empty string statement setting.
`,
			Switches: []string{
				cond_PROPER_NAME,
				cond_PROPER_DESCRIPTION,
				cond_PROPER_LOG,
				cond_PROPER_SWITCHES,
				cond_EMPTY_STRING_STATEMENT,
				expect_PANIC,
			},
		},
	}
}

func TestLogAPI(t *testing.T) {
	scenarios := _testLogScenarios()

	for i, s := range scenarios {
		s.ID = uint64(i)
		s.Name = suite_LOG_API

		// prepare
		ts := &Scenario{}
		testlib_ConfigureName(s, ts)
		testlib_ConfigureDescription(s, ts)
		testlib_ConfigureLog(s, ts)
		testlib_ConfigureSwitches(s, ts)

		statement := testlib_CreateStringStatement(s)

		// test
		_panick := Exec(func() any {
			if !HasCondition(s, cond_SUPPLY_NIL_SCENARIO) {
				Log(ts, statement)
			} else {
				Log(nil, statement)
			}
			return ""
		})
		panick := _panick.(string)

		// log output
		Log(s, _format("Test Scenario's ID		= %#v", ts.ID))
		Log(s, _format("Test Scenario's Name		= %#v", ts.Name))
		Log(s, _format("Test Scenario's Switches	= %#v", ts.Switches))
		Log(s, _format("Test Scenario's Logs		= %#v", ts.Logs))
		Log(s, _format("Got Panic			= %q", panick))

		// assert
		if !testlib_AssertPanic(s, panick) {
			Conclude(s, VERDICT_FAIL)
			t.Fail()
		}

		if !assert_Log(s, ts) {
			Conclude(s, VERDICT_FAIL)
			t.Fail()
		}

		if Conclusion(s) != VERDICT_FAIL {
			Conclude(s, VERDICT_PASS)
		}

		// report
		t.Logf("%s", ToString(s))
	}
}

func assert_Log(s, ts *Scenario) bool {
	if ts == nil {
		return HasCondition(s, cond_SUPPLY_NIL_SCENARIO)
	}
	// ts is now available

	if HasCondition(s, expect_PANIC) {
		switch {
		case HasCondition(s, cond_EMPTY_STRING_STATEMENT),
			HasCondition(s, cond_SUPPLY_NIL_SCENARIO):
			return true
		}
	}
	// no longer in panic expectation.

	if ts.Logs == nil {
		return false
	}
	// ts.Log is now available

	for _, v := range ts.Logs {
		switch v {
		case value_STATEMENT_1:
			return HasCondition(s, cond_PROPER_STRING_STATEMENT)
		default:
			continue
		}
	}

	return false
}
